keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -124,38 +124,71 @@ class EnvironmentPromotionChecker(
     deliveryConfig: DeliveryConfig,
     artifact: DeliveryArtifact,
     version: String,
-    targetEnvironment: String
+    targetEnvironment: Environment
   ) {
-    log.debug("Approving version $version of ${artifact.type} artifact ${artifact.name} in environment $targetEnvironment")
+    log.debug("Approving version $version of ${artifact.type} artifact ${artifact.name} in environment ${targetEnvironment.name}")
     val isNewVersion = repository
-      .approveVersionFor(deliveryConfig, artifact, version, targetEnvironment)
+      .approveVersionFor(deliveryConfig, artifact, version, targetEnvironment.name)
     if (isNewVersion) {
       log.info(
         "Approved {} {} version {} for {} environment {} in {}",
         artifact.name,
         artifact.type,
         version,
         deliveryConfig.name,
-        targetEnvironment,
+        targetEnvironment.name,
         deliveryConfig.application
       )
 
       publisher.publishEvent(
         ArtifactVersionApproved(
           deliveryConfig.application,
           deliveryConfig.name,
-          targetEnvironment,
+          targetEnvironment.name,
           artifact.name,
           artifact.type,
           version
         )
       )
 
+      // persist the status of stateless constraints because their current value is all we care about
+      freezeStatelessConstraintStatus(deliveryConfig, artifact, version, targetEnvironment)
+
       // recheck all resources in an environment, so action can be taken right away
-      repository.triggerResourceRecheck(targetEnvironment, deliveryConfig.application)
+      repository.triggerResourceRecheck(targetEnvironment.name, deliveryConfig.application)
     }
   }
 
+  /**
+   * Save the passing status of all stateless constraints when a version is approved so that
+   * their status stays the same forever. We don't want them to be evaluated anymore.
+   */
+  private fun freezeStatelessConstraintStatus(
+    deliveryConfig: DeliveryConfig,
+    artifact: DeliveryArtifact,
+    version: String,
+    environment: Environment
+  ) {
+    constraintRunner.generateStatelessConstraintState(
+      artifact = artifact,
+      deliveryConfig = deliveryConfig,
+      version = version,
+      environment = environment
+    ).forEach { constraintState ->
+      log.debug("Storing final constraint state for {} constraint for artifact {} {} version {} for {} environment {} in {}",
+        constraintState.type,
+        artifact.name,
+        artifact.type,
+        version,
+        deliveryConfig.name,
+        environment.name,
+        deliveryConfig.application
+      )
```

Comment:
```suggestion
      log.debug("Storing final state for constraint {} for {} version {} for {} environment {} in {}",
        constraintState.type,
        artifact,
        version,
        deliveryConfig.name,
        environment.name,
        deliveryConfig.application
      )
```
`DeliveryArtifact.toString()` already prints a nice representation of the artifact including the type.

Reply:
oh nice